templates/CMakeLists.txt.template

```diff
@@ -345,49 +370,53 @@
   #
-  #     protobuf_generate_grpc_cpp [<ARGN>...]
+  #     protobuf_generate_grpc_cpp <FILE_LOCATION> <IMPORT_PATH>
   #
-  #   ``ARGN``
-  #     ``.proto`` files
+  #   ``FILE_LOCATION``
+  #     The relative path of the ``.proto`` file to the project root
+  #   ``IMPORT_PATH``
+  #     The proto import path that itself expected to be placed in
   #
-  function(protobuf_generate_grpc_cpp)
-    if(NOT ARGN)
+  function(protobuf_generate_grpc_cpp FILE_LOCATION IMPORT_PATH)
+    if(NOT FILE_LOCATION)
       message(SEND_ERROR "Error: PROTOBUF_GENERATE_GRPC_CPP() called without any proto files")
       return()
     endif()
 
     set(_protobuf_include_path -I . -I <%text>${_gRPC_PROTOBUF_WELLKNOWN_INCLUDE_DIR}</%text>)
-    foreach(FIL <%text>${ARGN}</%text>)
-      get_filename_component(ABS_FIL <%text>${FIL}</%text> ABSOLUTE)
-      get_filename_component(FIL_WE <%text>${FIL}</%text> NAME_WE)
-      file(RELATIVE_PATH REL_FIL <%text>${CMAKE_CURRENT_SOURCE_DIR}</%text> <%text>${ABS_FIL}</%text>)
-      get_filename_component(REL_DIR <%text>${REL_FIL}</%text> DIRECTORY)
-      set(RELFIL_WE "<%text>${REL_DIR}/${FIL_WE}</%text>")
-
-      #if cross-compiling, find host plugin
-      if(CMAKE_CROSSCOMPILING)
-        find_program(_gRPC_CPP_PLUGIN grpc_cpp_plugin)
-      else()
-        set(_gRPC_CPP_PLUGIN $<TARGET_FILE:grpc_cpp_plugin>)
-      endif()
-
-      add_custom_command(
-        OUTPUT <%text>"${_gRPC_PROTO_GENS_DIR}/${RELFIL_WE}.grpc.pb.cc"</%text>
-               <%text>"${_gRPC_PROTO_GENS_DIR}/${RELFIL_WE}.grpc.pb.h"</%text>
-               <%text>"${_gRPC_PROTO_GENS_DIR}/${RELFIL_WE}_mock.grpc.pb.h"</%text>
-               <%text>"${_gRPC_PROTO_GENS_DIR}/${RELFIL_WE}.pb.cc"</%text>
-               <%text>"${_gRPC_PROTO_GENS_DIR}/${RELFIL_WE}.pb.h"</%text>
-        COMMAND <%text>${_gRPC_PROTOBUF_PROTOC_EXECUTABLE}</%text>
-        ARGS --grpc_out=<%text>generate_mock_code=true:${_gRPC_PROTO_GENS_DIR}</%text>
-             --cpp_out=<%text>${_gRPC_PROTO_GENS_DIR}</%text>
-             --plugin=protoc-gen-grpc=<%text>${_gRPC_CPP_PLUGIN}</%text>
-             <%text>${_protobuf_include_path}</%text>
-             <%text>${REL_FIL}</%text>
-        DEPENDS <%text>${ABS_FIL}</%text> <%text>${_gRPC_PROTOBUF_PROTOC}</%text> grpc_cpp_plugin
-        WORKING_DIRECTORY <%text>${CMAKE_CURRENT_SOURCE_DIR}</%text>
-        COMMENT "Running gRPC C++ protocol buffer compiler on <%text>${FIL}</%text>"
-        VERBATIM)
-    endforeach()
+    # compute the relative path
+    get_filename_component(ABS_FIL <%text>${_gRPC_PROTO_SRCS_DIR}/${IMPORT_PATH}</%text> ABSOLUTE)
+    get_filename_component(FIL_WE <%text>${_gRPC_PROTO_SRCS_DIR}/${IMPORT_PATH}</%text> NAME_WE)
+    file(RELATIVE_PATH REL_FIL <%text>${_gRPC_PROTO_SRCS_DIR}</%text> <%text>${ABS_FIL}</%text>)
+    get_filename_component(REL_DIR <%text>${REL_FIL}</%text> DIRECTORY)
+    set(RELFIL_WE "<%text>${REL_DIR}/${FIL_WE}</%text>")
+    # copy the proto file to a centralized location
```

Comment:
why do you need to copy the files to the a centralized location?
If this is because the .proto files need to see the .proto files they import, it seems that there would be a problem with dependencies of .proto files. Looks like you're copying proto files to a centralized location one by one, so by the time you copy A.proto (which depends on B.proto), B.proto might not be there yet?


If that's true, the order in which .proto files are listed in CMakeLists.txt is important and that seems pretty fragile (expecially because there's no easy what to influence the order in which they are listed).